tooling/test/sugar.go

```diff
@@ -208,6 +218,148 @@ func (e ExpectBuilder) BodyWithHint(hint string, body interface{}) ExpectBuilder
+			}
+		})
+	}
+}
+
+// Clone performs a deep clone of the ExpectBuilder
+// Note: if there are [check.Check]s used in the inner header or body components those are only shallowly cloned
+func (e ExpectBuilder) Clone() ExpectValidator {
+	clone := ExpectBuilder{}
+	var clonedHeaders []HeaderBuilder
+	for _, h := range e.Headers_ {
+		clonedHeaders = append(clonedHeaders, h.Clone())
+	}
+	clone.StatusCode_ = e.StatusCode_
+	clone.Headers_ = clonedHeaders
+
+	if e.Body_ == nil {
+		return clone
+	}
+
+	switch body := e.Body_.(type) {
+	case string:
+		clone.Body_ = body
+	case []byte:
+		clone.Body_ = body[:]
+	case check.CheckWithHint[string]:
+		clone.Body_ = body
+	case check.CheckWithHint[[]byte]:
+		clone.Body_ = body
+	case check.Check[string]:
+		clone.Body_ = body
+	case check.Check[[]byte]:
+		clone.Body_ = body
+	default:
+		panic("body must be string, []byte, or a regular check")
+	}
+	return clone
+}
+
+type AllOfExpectBuilder struct {
+	Expect_ []ExpectValidator `json:"expect,omitempty"`
+}
+
+var _ ExpectValidator = (*AllOfExpectBuilder)(nil)
+
+func AllOf(expect ...ExpectValidator) AllOfExpectBuilder {
+	return AllOfExpectBuilder{Expect_: expect}
+}
+
+func (e AllOfExpectBuilder) Validate(t *testing.T, res *http.Response, localReport Reporter) {
+	t.Helper()
+
+	for i, expect := range e.Expect_ {
+		t.Run(
+			fmt.Sprintf("Check %d", i), func(t *testing.T) {
+				expect.Validate(t, res, localReport)
+			})
+	}
+}
+
+// Clone performs a deep clone of the AllOfExpectBuilder
+// Note: if there are [check.Check]s used in the header or body components of the nested builders those are only
+// shallowly cloned
+func (e AllOfExpectBuilder) Clone() ExpectValidator {
+	var clonedInnerValidators []ExpectValidator
+	for _, eb := range e.Expect_ {
+		clonedInnerValidators = append(clonedInnerValidators, eb.Clone())
+	}
+	clone := AllOfExpectBuilder{Expect_: clonedInnerValidators}
+	return clone
+}
+
+type AnyOfExpectBuilder struct {
+	Expect_ []ExpectBuilder `json:"expect,omitempty"`
+}
+
+var _ ExpectValidator = (*AnyOfExpectBuilder)(nil)
+
+func AnyOf(expect ...ExpectBuilder) AnyOfExpectBuilder {
+	return AnyOfExpectBuilder{Expect_: expect}
+}
+
+func (e AnyOfExpectBuilder) Validate(t *testing.T, res *http.Response, localReport Reporter) {
+	t.Helper()
+
+	if len(e.Expect_) == 0 {
+		return
+	}
+
+	hadASuccessfulResponse := false
+	for i, expect := range e.Expect_ {
+		checks := validateResponse(t, expect, res)
+		responseSucceeded := true
+		for _, c := range checks {
+			if !c.checkOutput.Success {
+				responseSucceeded = false
+				break
+			}
+		}
+		if responseSucceeded {
+			hadASuccessfulResponse = true
+		}
+
+		t.Run(fmt.Sprintf("Check %d", i),
+			func(t *testing.T) {
+				if !responseSucceeded {
+					for _, c := range checks {
+						if c.checkOutput.Success {
+							t.Logf("Test %s passed", c.testName)
+						} else {
+							t.Logf("Test %s failed with: %s", c.testName, c.checkOutput.Reason)
+						}
```

Comment:
@laurentsenta this can be pretty noisy, I'm not sure if this is what you were looking for. Your suggestion https://github.com/ipfs/gateway-conformance/pull/162/files#r1322581089 implied a bunch of plumbing under the hood around `localReport` which I wasn't sure made sense to bundle in here as I'd like to get the basics here closed out.

If you think it's needed say so though and @hacdias or I can take a look

Reply:
Agreed, this is more complex than it needs to, afaii you'd need the recursivity on `Validate()`, and the side-effect on local report to make this a few loc. Let's keep it simple: feel free to drop that part, as long as the public API works for you, we'll revisit this later.